lib/storage/src/chunks/getChunkStream.ts

```diff
@@ -0,0 +1,43 @@
+import { RawDataPart } from "../Upload";
+import { Buffer } from "buffer";
+
+interface Buffers {
+  chunks: Buffer[];
+  length: number;
+}
+
+export async function* getChunk<T>(
+  data: T,
+  partSize: number,
+  getNextData: (data: T) => AsyncGenerator<Buffer>
+): AsyncGenerator<RawDataPart> {
+  let partNumber = 1;
+  const currentBuffer: Buffers = { chunks: [], length: 0 };
+
+  for await (const datum of getNextData(data)) {
+    currentBuffer.chunks.push(datum);
+    currentBuffer.length += datum.length;
+
+    while (currentBuffer.length >= partSize) {
+      // Concat all the buffers together once.
+      let dataChunk: Buffer = currentBuffer.chunks[0];
+      if(currentBuffer.chunks.length) {
+        dataChunk = Buffer.concat(currentBuffer.chunks);
+      }
```

Comment:
The chunks array is never going to be empty in this LoC as:
* The chunks and length is populated on lines 18 and 19. The while loop will not run if both `currentBuffer.chunks` and `datum` are emoty.
* The while loop enters only when `currentBuffer.length >= partSize`. Line 35 ensures that while loop exits in case currentBuffer.length was multiple of partSize (only case when `currentBuffer.chunks.length` can be empty)

If `currentBuffer.chunks` is empty, line 27 is going to throw `TypeError: Cannot read property 'slice' of undefined` any way.

```suggestion
      const dataChunk = Buffer.concat(currentBuffer.chunks);
```

Reply:
You're right that this should be `currentBuffer.chunks.length > 1`. But this branch is still needed. For example:

```
currentBuffer = {
    chunks: [.... length 55 ...]
    length: 55
} 
```

And we return chunks of size 10. We should only call concat if there are multiple chunks to concat. Otherwise we call concat 5 more times than needed on an array of size 1

Reply:
Sounds good!

How about using const instead of let like this with a comment about Buffer.concat?
```suggestion
      // Buffer.concat is slow operation and should be avoided for performance improvement.
      const dataChunk = currentBuffer.chunks.length > 1 ? Buffer.concat(currentBuffer.chunks) : currentBuffer.chunks[0];
```

Reply:
Added comment. Sure i'll update this to a one-line branch